sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedProducerAsyncClient.java

```diff
@@ -0,0 +1,334 @@
+                                    return Mono.empty();
+                                }
+                                return Mono.when(
+                                    producer.send(batch).doOnEach(signal -> {
+                                        if (signal.hasError()) {
+                                            this.clientOptions.getSendFailedContext()
+                                                .accept(new SendBatchFailedContext(batch.getEvents(), batch.getPartitionId(), signal.getThrowable()));
+                                        } else {
+                                            this.clientOptions.getSendSucceededContext()
+                                                .accept(new SendBatchSucceededContext(batch.getEvents(), batch.getPartitionId()));
+                                        }
+                                    }),
+                                    producer.createBatch(options).map(newBatch -> {
+                                        currentBatch.set(newBatch);
+                                        if (!newBatch.tryAdd(event)) {
+                                            this.clientOptions.getSendFailedContext()
+                                                .accept(
+                                                    new SendBatchFailedContext(
+                                                        eventDataBatch.getEvents(),
+                                                        eventDataBatch.getPartitionId(),
+                                                        new IllegalArgumentException(String.format(Locale.US, "Size of the payload exceeded maximum message size: %s kb", options.getMaximumSizeInBytes() / 1024)))
+                                                );
+                                        }
+                                        return newBatch;
+                                    })
+                                );
+                            });
+                        }).then().doFinally(sig -> {
+                            EventDataBatch batch = currentBatch.getAndSet(null);
+                            producer.send(batch).doOnEach(signal -> {
+                                if (signal.hasError()) {
+                                    this.clientOptions.getSendFailedContext()
+                                        .accept(new SendBatchFailedContext(batch.getEvents(), batch.getPartitionId(), signal.getThrowable()));
+                                } else {
+                                    this.clientOptions.getSendSucceededContext()
+                                        .accept(new SendBatchSucceededContext(batch.getEvents(), batch.getPartitionId()));
+                                }
+                            });
+                        });
+                    })
+                    .subscribeOn(Schedulers.newSingle(partitionId))
+                    .subscribe();
```

Comment:
`subscribe` is used at the end of your reactive chain. It shouldn't be used here because all it does is kick-off the work, but it doesn't wait for the work to complete (ie. fire and forget.) Similar to TypeScript/JS promises, where you can chain promises. You want to chain these async operations together.

Reply:
I'd like to make this code clear. 

Actually kicking off the work and don't wait for complete is what I want. User can just enqueue events without waiting for result. Because we set `SendBatchSucceededContext` and `SendBatchFailedContext` to hanlde the result. The reason why I generate a Flux running on another thread by using `publishOn` and `subscribeOn` for a long term is I can use `windowTimeout` to implement the feature of `maxPendingEventCount` and `maxWaitTime` by `take` event from blockingQueue and publish into `windowTimeout`. Otherwise, I need to implement a state machine to record the time that events enqueue. I have no idea to put events from multi threads into a same Flux.

There is one thing. Actually I generate a Flux for each partition. If there are thousand partitions, I'm not sure the memory usage looks like.